server/aws-lsp-codewhisperer/src/language-server/chat/chatController.ts

```diff
@@ -200,24 +200,42 @@ export class ChatController implements ChatHandlers {
             return
         }
 
-        const cursorPosition = params.cursorPosition
+        let cursorPosition = params.cursorPosition
 
         const indentRange = {
             start: { line: cursorPosition.line, character: 0 },
             end: cursorPosition,
         }
         const documentContent = await this.#features.workspace.getTextDocument(params.textDocument.uri)
         let indent = documentContent?.getText(indentRange)
+
+        let hasVirtualSpace = false
         if (indent && indent.trim().length !== 0) {
             indent = ' '.repeat(indent.length - indent.trimStart().length)
+        } else if ((isNullish(indent) || !indent) && cursorPosition.character > 0) {
+            // When text content at indent range is undefined or empty but the cursor is not at position 0
+            // It means there are virtual spaces that is being rendered by the IDE
+            // In this case, the indentation is determined by the cursorPosition
+            this.#log('Indent is nullish and the cursor position is greater than zero while inserting code')
+            indent = ' '.repeat(cursorPosition.character)
+            hasVirtualSpace = true
+            cursorPosition.character = 0
         }
 
         let textWithIndent = ''
         params.code.split('\n').forEach((line, index) => {
             if (index === 0) {
+                if (hasVirtualSpace) {
+                    textWithIndent += indent
```

Comment:
Similar to Line 233, if the first line is blank, it shouldn't receive indentations. (Can this also be added as a test case please?)

Reply:
I was under the impression that in the case of a virtual space we actually need this extra indent. 

For a regular insertion (no virtual spaces) we don't add any indentation on the first line since the expectation is to insert the code block at the current cursor position and adding an indentation to the insertion would break the expectation

However, in the case of virtual spaces, the cursor might appear to the user at a certain position due to virtual spaces (e.g., 4 virtual spaces on a line, with the server receiving the cursor position as character 4). In this situation, the server must convert the virtual indentation into a real indentation when inserting the content into the file.

This is what the line in question achieves. To ensure the indentation is applied correctly, the server adjusts the cursor position to character 0 on line 222, so that the real spaces are inserted at the beginning of the line.

Reply:
Here is an arbitrary example for illustrative purposes. My comment above refers to line 81 in the sample image, but I have provided a complete representation of the desired outcome. It's a corner case where the first line is blank - generally this should be uncommon.
![image](https://github.com/user-attachments/assets/91890d98-2533-4374-a0b1-6e51e0ec1431)

Reply:
Good catch and a nice demonstration, this makes sense to me. I will update the code and continue on adding the test cases